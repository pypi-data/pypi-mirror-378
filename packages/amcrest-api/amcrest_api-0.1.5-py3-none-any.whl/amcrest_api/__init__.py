"""
API Wrapper for Amcrest V3.26

https://support.amcrest.com/hc/en-us/articles/17903073032973-Amcrest-HTTP-API-SDK
"""

from importlib import metadata as importlib_metadata


def get_version() -> str:
    """Version."""
    try:
        return importlib_metadata.version(__name__)
    except importlib_metadata.PackageNotFoundError:  # pragma: no cover
        return "unknown"


version: str = get_version()
