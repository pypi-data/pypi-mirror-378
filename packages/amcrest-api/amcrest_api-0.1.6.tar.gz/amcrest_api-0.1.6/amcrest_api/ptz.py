"""PTZ Helpers."""

from __future__ import annotations

from dataclasses import KW_ONLY, dataclass, field
from enum import StrEnum
from typing import Any


@dataclass(frozen=True)
class PtzPresetData:
    """Data for PTZ preset."""

    index: int
    name: str


class Direction(StrEnum):
    """Directions."""

    UP = "UP"
    DOWN = "DOWN"
    LEFT = "LEFT"
    RIGHT = "RIGHT"
    LEFTUP = "LEFTUP"
    RIGHTUP = "RIGHTUP"
    LEFTDOWN = "LEFTDOWN"
    RIGHTDOWN = "RIGHTDOWN"


ALL_DIRECTIONS = set(Direction)


@dataclass(kw_only=True)
class PtzCapabilityData:
    """Data for PTZ capabilities."""

    pan: bool = False
    tilt: bool = False
    zoom: bool = False
    preset: bool = False
    tour: bool = False

    pan_min: float = 0.0
    pan_max: float = 0.0
    pan_speed_min: int = 1
    pan_speed_max: int = 1

    tilt_min: float = 0.0
    tilt_max: float = 0.0
    tilt_speed_min: int = 1
    tilt_speed_max: int = 1

    zoom_min: float = 1.0
    zoom_max: float = 1.0

    preset_min: int = 0
    preset_max: int = 0

    tour_min: int = 0
    tour_max: int = 0

    supported_directions: list[Direction] = field(default_factory=list[Direction])

    @staticmethod
    def create_from_response(response: dict[str, Any]) -> PtzCapabilityData:
        """Create from a PTZ capabilities response."""
        caps: dict[str, Any] = response["caps"]

        def check_true(key: str) -> bool:
            if (cap_str := caps.get(key)) is not None:
                return cap_str.lower() == "true"
            return False

        ret = PtzCapabilityData(
            pan=check_true("Pan"),
            tilt=check_true("Tile"),  # the key is a typo in the API
            zoom=check_true("Zoom"),
            preset=check_true("Preset"),
            tour=check_true("Tour"),
        )
        if ret.pan:
            ret.pan_min = float(caps["PtzMotionRange"]["HorizontalAngle"][0])
            ret.pan_max = float(caps["PtzMotionRange"]["HorizontalAngle"][1])
            ret.pan_speed_min = int(caps["PanSpeedMin"])
            ret.pan_speed_max = int(caps["PanSpeedMax"])
            if ret.pan_min > ret.pan_max:
                ret.pan_min, ret.pan_max = ret.pan_max, ret.pan_min
        if ret.tilt:
            ret.tilt_min = float(caps["PtzMotionRange"]["VerticalAngle"][0])
            ret.tilt_max = float(caps["PtzMotionRange"]["VerticalAngle"][1])
            ret.pan_speed_min = int(caps["PanSpeedMin"])
            ret.pan_speed_max = int(caps["PanSpeedMax"])
            if ret.tilt_min > ret.tilt_max:
                ret.tilt_min, ret.tilt_max = ret.tilt_max, ret.tilt_min
        if ret.zoom:
            ret.zoom_min = float(caps["ZoomMin"])
            ret.zoom_max = float(caps["ZoomMax"])
        if ret.preset:
            ret.preset_min = int(caps["PresetMin"])
            ret.preset_max = int(caps["PresetMax"])
        if ret.tour:
            ret.tour_min = int(caps["TourMin"])
            ret.tour_max = int(caps["TourMax"])
        if ret.pan or ret.tilt:
            ret.supported_directions = sorted(
                list[Direction](
                    ALL_DIRECTIONS
                    ^ set[Direction](caps.get("UnSupportDirections", dict()).values())  # type: ignore
                )
            )
        return ret


@dataclass(kw_only=True)
class PtzStatusData:
    """Data for PTZ status."""

    action: str | None
    move_status: str
    zoom_status: str
    preset_id: int | None = None
    sequence: int | None = None
    position_pan: float
    position_tilt: float
    position_zoom: float
    pts: int | None
    utc: int | None

    @staticmethod
    def create_from_response(response: dict[str, Any]) -> PtzStatusData:
        """Create from a PTZ Status response."""
        ptz_status_dict: dict[str, Any] = response["status"]  # expect nested dict
        if ptz_status_dict.get("PresetID") == "0":
            preset_id = None
        else:
            preset_id = int(ptz_status_dict["PresetID"])
        if ptz_status_dict.get("Sequence") == "0":
            sequence = None
        else:
            sequence = int(ptz_status_dict["Sequence"])
        return PtzStatusData(
            action=ptz_status_dict.get("Action"),
            move_status=ptz_status_dict["MoveStatus"],
            zoom_status=ptz_status_dict["ZoomStatus"],
            preset_id=preset_id,
            sequence=sequence,
            # Typo in API "Postion instead of Position" # codespell:ignore
            position_pan=float(ptz_status_dict["Postion"][0]),  # codespell:ignore
            position_tilt=float(ptz_status_dict["Postion"][1]),  # codespell:ignore
            position_zoom=float(ptz_status_dict["Postion"][2]),  # codespell:ignore
            pts=int(str(ptz_status_dict.get("PTS"))),
            utc=int(str(ptz_status_dict.get("UTC"))),
        )


@dataclass(kw_only=True)
class PtzRelativeMove:
    """Class for a relative move. All values are normalized from [-1,1]."""

    _caps: PtzCapabilityData | None = None
    _horizontal: float | None = None
    _vertical: float | None = None
    _zoom: float | None = None
    _channel: int = 1

    def __init__(
        self,
        *,
        caps: PtzCapabilityData | None = None,
        horizontal: float | None = None,
        vertical: float | None = None,
        zoom: float | None = None,
        channel: int = 1,
    ):
        self.caps = caps
        self.horizontal = horizontal
        self.vertical = vertical
        self.zoom = zoom
        self.channel = channel

    @property
    def caps(self) -> PtzCapabilityData | None:
        """PTZ capabilities."""
        return self._caps

    @caps.setter
    def caps(self, value: PtzCapabilityData | None):
        self._caps = value

    @property
    def horizontal(self) -> float | None:
        """Relative horizontal movement, [-1,1] inclusive"""
        return self._horizontal

    @horizontal.setter
    def horizontal(self, value: float | None) -> None:
        if value is None or -1.0 <= value <= 1.0:
            self._horizontal = value
        else:
            raise ValueError("Must be between -1.0 and 1.0 inclusive")

    @property
    def vertical(self) -> float | None:
        """Relative vertical movement, [-1,1] inclusive"""
        return self._vertical

    @vertical.setter
    def vertical(self, value: float | None) -> None:
        if value is None or -1.0 <= value <= 1.0:
            self._vertical = value
        else:
            raise ValueError("Must be between -1.0 and 1.0 inclusive")

    @property
    def zoom(self) -> float | None:
        """Relative zoom movement, [-1,1] inclusive"""
        return self._zoom

    @zoom.setter
    def zoom(self, value: float | None) -> None:
        if value is None or -1.0 <= value <= 1.0:
            self._zoom = value
        else:
            raise ValueError("Must be between -1.0 and 1.0 inclusive")

    @property
    def channel(self) -> int:
        """PTZ channel."""
        return self._channel

    @channel.setter
    def channel(self, value: int) -> None:
        self._channel = int(value)

    def get_query_dict(self) -> dict[str, Any]:
        """
        Get a dictionary that can be passed as query parameters to the API Endpoint.

        arg1 is horizontal
        arg2 is vertical
        arg3 is zoom
        """
        ret: dict[str, Any] = {
            "action": "moveRelatively",
            "channel": self.channel,
            "arg1": 0,
            "arg2": 0,
            "arg3": 0,
        }
        if self.horizontal is not None:
            ret["arg1"] = self.horizontal
        if self.vertical is not None:
            ret["arg2"] = self.vertical
        if self.zoom is not None:
            ret["arg3"] = self.zoom
        return ret


@dataclass
class PtzBasicMove:
    """Class for starting PTZ motion."""

    direction: Direction
    _: KW_ONLY
    caps: PtzCapabilityData | None = None
    speed: int = 1
    vertical_speed: int | None = None
    horizontal_speed: int | None = None
    channel: int = 1

    def get_start_query_dict(self) -> dict[str, Any]:
        """
        Get a dictionary that can be passed as query parameters to the API Endpoint.

        arg1 is vertical speed on combo moves
        arg2 is vertical or horizontal speed on single moves
        arg3 is 0
        """
        ret: dict[str, Any] = {
            "action": "start",
            "channel": self.channel,
            "arg1": 0,
            "arg2": 0,
            "arg3": 0,
        }
        if self.direction in [
            Direction.LEFTDOWN,
            Direction.RIGHTDOWN,
            Direction.RIGHTUP,
            Direction.LEFTUP,
        ]:
            ret["arg1"] = self.vertical_speed or self.speed
            ret["arg2"] = self.horizontal_speed or self.speed

        elif self.direction in [
            Direction.LEFT,
            Direction.RIGHT,
            Direction.UP,
            Direction.DOWN,
        ]:
            ret["arg1"] = 0
            ret["arg2"] = self.speed
        ret["code"] = self.direction.title()
        ret["arg3"] = 0
        return ret

    def get_stop_query_dict(self) -> dict[str, Any]:
        """Get the stop query dict."""
        return {
            "action": "stop",
            "channel": self.channel,
            "code": self.direction.title(),
            "arg1": 0,
            "arg2": 0,
            "arg3": 0,
        }


@dataclass(kw_only=True)
class PtzAccuratePosition:
    """Class for Accurate position setting."""

    caps: PtzCapabilityData | None = None
    horizontal_position: float | None = None
    vertical_position: float | None = None
    zoom: float | None = None
    channel: int = 1

    def get_query_dict(self) -> dict[str, Any]:
        """
        Get a dictionary to be passed as query parameters to the Accurate Position API

        arg1 horizonatal position
        arg2 vertical position
        arg3 zoom
        """
        if self.caps is None:
            raise RuntimeError(
                "Cannot get query dict unless PTZ capabilities are assigned."
            )

        ret: dict[str, Any] = {
            "action": "moveAbsolutely",
            "channel": self.channel,
            "arg1": 0,
            "arg2": 0,
            "arg3": 0,
        }
        if self.horizontal_position is not None and self.caps.pan:
            if self.caps.pan_min <= self.horizontal_position <= self.caps.pan_max:
                ret["arg1"] = (self.horizontal_position - 180.0) / 180.0
            else:
                raise ValueError(
                    f"Horizontal move degrees must be in the range [{self.caps.pan_min}, {self.caps.pan_max}, but it was {self.horizontal_position}]"  # noqa: E501
                )
        if self.vertical_position is not None and self.caps.tilt:
            if self.caps.tilt_min <= self.vertical_position <= self.caps.tilt_max:
                ret["arg2"] = self.vertical_position / 180.0
            else:
                raise ValueError(
                    f"Vertical move degrees must be in the range [{self.caps.tilt_min}, {self.caps.tilt_max}, but it was {self.vertical_position}]"  # noqa: E501
                )
        if self.zoom is not None and self.caps.zoom:
            if self.caps.zoom_min <= self.zoom <= self.caps.zoom_max:
                zoom_range = self.caps.zoom_max - self.caps.zoom_min
                zoom_mid = (self.caps.zoom_max + self.caps.zoom_min) / 2.0
                normed = (self.zoom - zoom_mid) / (zoom_range / 2)
                ret["arg3"] = normed
            else:
                raise ValueError(
                    f"Zoom must be in the range [{self.caps.zoom_min}, {self.caps.zoom_max}, but it was {self.zoom}]"  # noqa: E501
                )
        return ret
