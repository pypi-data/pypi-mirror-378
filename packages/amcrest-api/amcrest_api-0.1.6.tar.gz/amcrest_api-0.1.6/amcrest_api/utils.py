"""Utilities."""

import re
from typing import Any

import httpx


def parse_response(response: httpx.Response) -> bytes | dict:
    content_type: str = response.headers["content-type"]
    if "image/jpeg" in content_type:
        return response.content
    elif "text/plain" in content_type:
        return parse_key_value_response(response)
    elif "application/json" in content_type:
        return response.json()
    raise ValueError(f"Response does not contain supported content: {content_type}")


def parse_key_value_response(response: httpx.Response) -> dict[int | str, Any]:
    ret: dict[int | str, Any] = {}
    # The API can simply reply with OK too
    if response.text.strip() == "OK":
        return ret
    for line in response.iter_lines():
        keystr, val = line.split("=")
        keysraw = keystr.split(".")
        if keysraw[0] == "table":
            keysraw = keysraw[1:]
        keys: list[list[int | str]] = []
        for keyraw in keysraw:
            mainkey = keyraw.split("[")[0]  # get the key without indices
            indices = [int(x) for x in re.findall(r"\[(\d+)\]", keyraw)]
            keys.append([mainkey] + indices)
        # flatten the keys
        keysflat: list[int | str] = [x for xx in keys for x in xx]
        current = ret
        for keyflat in keysflat[:-1]:
            current = current.setdefault(keyflat, {})
        current[keysflat[-1]] = val

    return ret


def indexed_dict_to_list(indexed_dict: dict[int, Any]) -> list[Any]:
    """Converts a dict that could be a list, into a list."""
    return [indexed_dict[i] for i, key in enumerate(sorted(indexed_dict.keys()))]
