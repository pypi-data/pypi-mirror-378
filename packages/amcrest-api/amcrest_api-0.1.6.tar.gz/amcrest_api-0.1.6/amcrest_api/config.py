"""Config for information unlikely to change during a session."""

from dataclasses import dataclass
from typing import Any

from amcrest_api.const import StreamType, StreamTypeName

from .event import EventMessageType
from .ptz import PtzCapabilityData


@dataclass(frozen=True)
class Config:
    """A Config containing elements unlikely to change during the API Session."""

    device_type: str
    hardware_version: str
    machine_name: str
    max_extra_stream: StreamType
    network: dict[str, Any]
    ptz_capabilities: PtzCapabilityData
    serial_number: str
    session_physical_address: str
    supported_events: list[EventMessageType]
    supported_streams: dict[StreamType, StreamTypeName]
    software_version: str
