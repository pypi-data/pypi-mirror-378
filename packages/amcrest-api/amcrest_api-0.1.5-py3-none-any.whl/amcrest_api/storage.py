"""Storage helpers."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .utils import indexed_dict_to_list


@dataclass(frozen=True, kw_only=True)
class StorageDeviceInfo:
    """Storage device info dataclass."""

    is_error: bool
    path: str
    type: str
    cant_hot_plug: bool
    total_bytes: int
    used_bytes: int
    life_percent: float
    state: str
    sd_encrypt_flag: int
    health_data_flag: int
    name: str

    @staticmethod
    def create_from_response(response: dict[str, Any]) -> list[StorageDeviceInfo]:
        """Create from a PTZ capabilities response."""
        infos: list[dict[str, Any]] = indexed_dict_to_list(response["list"]["info"])

        ret: list[StorageDeviceInfo] = []

        for i, info in enumerate(infos):
            detail = info["Detail"][i]
            ret.append(
                StorageDeviceInfo(
                    is_error=detail["IsError"] == "true",
                    path=detail["Path"],
                    type=detail["Type"],
                    cant_hot_plug=info.get("CantHotPlug", "false") == "true",
                    total_bytes=round(float(detail["TotalBytes"])),
                    used_bytes=round(float(detail["UsedBytes"])),
                    life_percent=float(info["LifePercent"]),
                    state=info["State"],
                    sd_encrypt_flag=int(info["SDEncryptFlag"]),
                    health_data_flag=int(info["HealthDataFlag"]),
                    name=info["Name"],
                )
            )

        return ret
