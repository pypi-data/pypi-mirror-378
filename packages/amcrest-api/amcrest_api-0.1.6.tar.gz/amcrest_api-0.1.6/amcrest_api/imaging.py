"""Classes for camra imaging."""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum, IntFlag, StrEnum
from typing import Any

from .utils import indexed_dict_to_list


class Rotate90Flag(IntFlag):
    """Rotation flags."""

    NO_ROTATE = 0
    CLOCKWISE_90 = 1
    COUNTERCLOCKWISE_90 = 2


@dataclass(kw_only=True)
class VideoImageControl:
    """Video Image Control (Rotation/Flip)"""

    flip: bool = False
    freeze: bool = False
    mirror: bool = False
    rotate_90: Rotate90Flag = Rotate90Flag.NO_ROTATE
    stable: int = 0

    @staticmethod
    def create_from_response(response: dict[str, Any]) -> list[VideoImageControl]:
        """Create from a Video Image Control response."""
        return [
            VideoImageControl(
                flip=control.get("Flip", "false") == "true",
                freeze=control.get("Freeze", "false") == "true",
                mirror=control.get("Mirror", "false") == "true",
                rotate_90=Rotate90Flag(int(control.get("Rotate90", "0"))),
                stable=int(control.get("Stable", "0")),
            )
            for control in indexed_dict_to_list(response["VideoImageControl"])
        ]


class VideoMode(StrEnum):
    """Selections for video modes."""

    COLOR = "Color"
    BRIGHTNESS = "Brightness"
    BLACK_WHITE = "BlackWhite"
    PHOTORESISTOR = "Photoresistor"
    GAIN = "Gain"


class Sensitivity(IntEnum):
    """Sensitivity settings."""

    LOW = 1
    MEDIUM = 2
    HIGH = 3


class ConfigNo(IntEnum):
    """Video Profile Config Index"""

    DAY = 0
    NIGHT = 1
    NORMAL = 2


CONFIG_NO_DICT = {
    ConfigNo.DAY: "Day",
    ConfigNo.NIGHT: "Night",
    ConfigNo.NORMAL: "Normal/General",
}


class VideoDayNightType(StrEnum):
    """Video Day/Night Types"""

    ELECTRON = "Electron"
    MECHANISM = "Mechanism"
    NIGHT_ICR = "NightICR"
    AUTO = "Auto"


@dataclass(kw_only=True)
class VideoDayNight:
    """Video Day/Night Settings"""

    delay_seconds: int
    mode: VideoMode
    sensitivity: Sensitivity
    type: VideoDayNightType

    @staticmethod
    def create_from_response(
        response: dict[str, Any],
    ) -> list[list[VideoDayNight]]:
        """Create from API response."""
        return [
            [
                VideoDayNight(
                    delay_seconds=int(config["Delay"]),
                    mode=VideoMode(config["Mode"]),
                    sensitivity=Sensitivity(int(config["Sensitivity"])),
                    type=VideoDayNightType(config["Type"]),
                )
                for config in indexed_dict_to_list(video_in_day_night)
            ]
            for video_in_day_night in indexed_dict_to_list(response["VideoInDayNight"])
        ]


class LightingMode(StrEnum):
    """Lighting Modes."""

    MANUAL = "Manual"
    AUTO = "Auto"
    OFF = "Off"


@dataclass(kw_only=True)
class Light:
    """Light."""

    angle: int = 50
    light: int


@dataclass(kw_only=True)
class Lighting:
    """Lighting config."""

    mode: LightingMode
    correction: int | None = None
    # far_light: int
    middle_light: Light | None
    # near_light: int
    light_type: str | None = None
    sensitivity: Sensitivity | None = None

    @staticmethod
    def create_from_response(
        response: dict[str, Any],
    ) -> list[list[list[Lighting]]]:
        """Create from API response."""
        return [
            [
                [
                    Lighting(
                        mode=light["Mode"],
                        correction=light["Correction"],
                        # far_light=int(light["FarLight"][0]["Light"]),
                        middle_light=Light(
                            light=int(light["MiddleLight"][0]["Light"]),
                            angle=int(light["MiddleLight"][0].get("Angle", 50)),
                        ),
                        light_type=light.get("LightType"),
                        # near_light=int(light["NearLight"][0]["Light"]),
                        sensitivity=Sensitivity(int(light["Sensitive"])),
                    )
                    for light in indexed_dict_to_list(config)
                ]
                for config in indexed_dict_to_list(lighting)
            ]
            for lighting in indexed_dict_to_list(response["Lighting_V2"])
        ]
