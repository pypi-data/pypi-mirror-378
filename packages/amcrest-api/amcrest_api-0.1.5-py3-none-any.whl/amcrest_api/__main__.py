"""Main"""

import asyncio
from enum import Enum
from functools import cached_property
from pprint import pprint

import typer
from rich.console import Console
from sshkeyboard import listen_keyboard

from . import version
from .camera import Camera
from .imaging import ConfigNo, Rotate90Flag, VideoMode
from .ptz import PtzPresetData, PtzRelativeMove


class Color(str, Enum):
    white = "white"
    red = "red"
    cyan = "cyan"
    magenta = "magenta"
    yellow = "yellow"
    green = "green"


app = typer.Typer(
    name="amcrest-api",
    help="API Wrapper for Amcrest V3.26",
    add_completion=False,
)
console = Console()

POSSIBLE_ACTIONS = [
    attr
    for attr, value in vars(Camera).items()
    if isinstance(value, (cached_property, property))
]


def version_callback(print_version: bool) -> None:
    """Print the version of the package."""
    if print_version:
        print(version)
        raise typer.Exit()


@app.command(name="")
def main(
    print_version: bool = typer.Option(
        None,
        "-v",
        "--version",
        callback=version_callback,
        is_eager=True,
        help="Prints the version of the amcrest-api package.",
    ),
    username: str = typer.Option(
        ...,
        "-u",
        "--username",
        envvar="AMC_USERNAME",
        case_sensitive=True,
        help="User name for Amcrest Camera",
    ),
    password: str = typer.Option(
        ...,
        "-p",
        "--password",
        envvar="AMC_PASSWORD",
        case_sensitive=True,
        help="Password for Amcrest Camera",
    ),
    host: str = typer.Option(
        ...,
        "-h",
        "--host",
        envvar="AMC_HOST",
        case_sensitive=False,
        help="Host name or IP Address for Amcrest Camera",
    ),
) -> None:
    """Print a greeting with a giving name."""

    cam = Camera(host=host, username=username, password=password)

    async def print_config(cam: Camera):
        pprint(await cam.async_get_fixed_config())
        pprint(await cam.async_get_current_time())

    asyncio.run(print_config(cam))

    print("Listening for keypresses...")

    async def on_press(key):
        try:
            if key == "up":
                await cam.async_ptz_move_relative(PtzRelativeMove(vertical=0.1))
            elif key == "down":
                await cam.async_ptz_move_relative(PtzRelativeMove(vertical=-0.1))
            elif key == "left":
                await cam.async_ptz_move_relative(PtzRelativeMove(horizontal=-0.05))
            elif key == "right":
                await cam.async_ptz_move_relative(PtzRelativeMove(horizontal=0.05))

            if key in ["pageup", "pagedown"]:
                video_image_control = (await cam.async_video_image_control)[0]

                if key == "pageup":
                    if (
                        video_image_control.rotate_90 == Rotate90Flag.NO_ROTATE
                        and not video_image_control.flip
                    ):
                        video_image_control.rotate_90 = Rotate90Flag.CLOCKWISE_90
                    elif (
                        video_image_control.rotate_90 == Rotate90Flag.CLOCKWISE_90
                        and not video_image_control.flip
                    ):
                        video_image_control.flip = True
                        video_image_control.rotate_90 = Rotate90Flag.NO_ROTATE
                    elif (
                        video_image_control.rotate_90 == Rotate90Flag.NO_ROTATE
                        and video_image_control.flip
                    ):
                        video_image_control.flip = False
                        video_image_control.rotate_90 = Rotate90Flag.COUNTERCLOCKWISE_90
                    else:
                        video_image_control.flip = False
                        video_image_control.rotate_90 = Rotate90Flag.NO_ROTATE

                elif key == "pagedown":
                    if (
                        video_image_control.rotate_90 == Rotate90Flag.NO_ROTATE
                        and not video_image_control.flip
                    ):
                        video_image_control.rotate_90 = Rotate90Flag.COUNTERCLOCKWISE_90
                    elif (
                        video_image_control.rotate_90 == Rotate90Flag.CLOCKWISE_90
                        and not video_image_control.flip
                    ):
                        video_image_control.rotate_90 = Rotate90Flag.NO_ROTATE
                    elif (
                        video_image_control.rotate_90 == Rotate90Flag.NO_ROTATE
                        and video_image_control.flip
                    ):
                        video_image_control.flip = False
                        video_image_control.rotate_90 = Rotate90Flag.CLOCKWISE_90
                    else:
                        video_image_control.flip = True
                        video_image_control.rotate_90 = Rotate90Flag.NO_ROTATE
                await cam.async_set_video_image_control(video_image_control)

            SET_PRESET_MAP = {
                "!": "1",
                "@": "2",
                "#": "3",
                "$": "4",
                "%": "5",
                "^": "6",
                "&": "7",
                "*": "8",
            }
            if key in SET_PRESET_MAP.values():
                await cam.async_ptz_move_to_preset(int(key))

            if key in SET_PRESET_MAP:
                preset = PtzPresetData(
                    SET_PRESET_MAP[key], f"MyCustomPreset{SET_PRESET_MAP[key]}"
                )
                await cam.async_set_ptz_preset(preset)
                print(f"saved prest: {preset}")

            if key in "x":
                for preset in await cam.async_ptz_preset_info:
                    await cam.async_clear_ptz_preset(preset)
                    print(f"cleared preset: {preset}")

            if key == "r":
                pprint(await cam.async_ptz_preset_info)

            if key == "space":
                pm = await cam.async_get_privacy_mode_on()
                await cam.async_set_privacy_mode_on(not pm)

            if key == "d":
                pprint(await cam.async_get_video_in_day_night())

            if key == "D":
                config = (await cam.async_get_video_in_day_night())[0][ConfigNo.NORMAL]
                if config.mode == VideoMode.BRIGHTNESS:
                    config.mode = VideoMode.BLACK_WHITE
                elif config.mode == VideoMode.BLACK_WHITE:
                    config.mode = VideoMode.COLOR
                elif config.mode == VideoMode.COLOR:
                    config.mode = VideoMode.BRIGHTNESS
                await cam.async_set_video_in_day_night(config, ConfigNo.NORMAL)
                print(f"Switched to mode: {config.mode}")

            if key == "l":
                lights = await cam.async_lighting_config
                pprint(lights)

            if key == "c":
                pprint(await cam.async_get_fixed_config())

            if key == "s":
                pprint(await cam.async_storage_info)

            if key == "#":
                pprint(f"Extra Streams: {await cam.async_max_extra_stream}")

            if key == "m":
                try:
                    async for event in cam.async_listen_events():
                        print(event)
                except KeyboardInterrupt:
                    pass
        except Exception as e:
            print(type(e), e)

    listen_keyboard(on_press=on_press, lower=False)


if __name__ == "__main__":
    app()
