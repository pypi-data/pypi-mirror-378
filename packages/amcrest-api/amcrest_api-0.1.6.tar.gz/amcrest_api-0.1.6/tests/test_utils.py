"""Test utils.py."""

import pytest

from amcrest_api import utils


def test_parse_json(mock_json_response, snapshot):
    """Test json responses are parsed."""
    res = utils.parse_response(response=mock_json_response)
    assert isinstance(res, dict)
    assert res == snapshot


def test_parse_key_value_table(mock_key_value_with_table_response, snapshot):
    """Test key value responses with tables are parsed."""
    res = utils.parse_response(response=mock_key_value_with_table_response)
    assert isinstance(res, dict)
    # spot check
    assert res["General"]["LockLoginTimes"] == "30"
    # snapshot check
    assert res == snapshot


def test_parse_key_value_with_array(mock_key_value_with_array_response, snapshot):
    """Test key value responses with arrays are parsed."""
    res = utils.parse_response(response=mock_key_value_with_array_response)
    assert isinstance(res, dict)
    # spot  check
    assert res["Snap"][0]["HolidayEnable"] == "false"
    assert res["Snap"][0]["TimeSection"][0][1] == "0 00:00:00-23:59:59"
    # snapshot check
    assert res == snapshot


def test_parse_single_key_value(mock_key_value_response):
    """Test a response with a single key-value"""
    res = utils.parse_response(response=mock_key_value_response)
    assert res["sn"] == "AMC0"


def test_indexed_dict_to_list():
    """Test converting a dict of contiguous indices can be converted to a list."""
    indexed_dict = {
        2: "StorageLowSpace",
        3: "StorageNotExist",
        0: "AudioAnomaly",
        1: "StorageFailure",
        4: "VideoBlind",
        5: "VideoMotion",
    }
    assert utils.indexed_dict_to_list(indexed_dict) == [
        "AudioAnomaly",
        "StorageFailure",
        "StorageLowSpace",
        "StorageNotExist",
        "VideoBlind",
        "VideoMotion",
    ]

    # fails if not all in a sequence
    indexed_dict.pop(3)
    with pytest.raises(KeyError):
        assert utils.indexed_dict_to_list(indexed_dict)
