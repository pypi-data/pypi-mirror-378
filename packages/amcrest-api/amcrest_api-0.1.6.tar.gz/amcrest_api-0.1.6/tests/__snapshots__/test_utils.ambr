# serializer version: 1
# name: test_parse_json
  dict({
    'proxy': dict({
      'IP': '10.1.2.3',
      'Port': 8080,
    }),
    'way': 0,
  })
# ---
# name: test_parse_key_value_table
  dict({
    'General': dict({
      'LocalNo': '8',
      'LockLoginEnable': 'true',
      'LockLoginTimes': '30',
      'LoginFailLockTime': '300',
      'MachineName': 'AMC05820_4FEB51',
      'MaxOnlineTime': '3600',
    }),
  })
# ---
# name: test_parse_key_value_with_array
  dict({
    'Snap': dict({
      0: dict({
        'HolidayEnable': 'false',
        'TimeSection': dict({
          0: dict({
            0: '6 00:00:00-23:59:59',
            1: '0 00:00:00-23:59:59',
            2: '0 00:00:00-23:59:59',
            3: '0 00:00:00-23:59:59',
            4: '0 00:00:00-23:59:59',
            5: '0 00:00:00-23:59:59',
          }),
          1: dict({
            0: '6 00:00:00-23:59:59',
            1: '0 00:00:00-23:59:59',
            2: '0 00:00:00-23:59:59',
            3: '0 00:00:00-23:59:59',
            4: '0 00:00:00-23:59:59',
            5: '0 00:00:00-23:59:59',
          }),
          2: dict({
            0: '6 00:00:00-23:59:59',
            1: '0 00:00:00-23:59:59',
            2: '0 00:00:00-23:59:59',
            3: '0 00:00:00-23:59:59',
            4: '0 00:00:00-23:59:59',
            5: '0 00:00:00-23:59:59',
          }),
          3: dict({
            0: '6 00:00:00-23:59:59',
            1: '0 00:00:00-23:59:59',
            2: '0 00:00:00-23:59:59',
            3: '0 00:00:00-23:59:59',
            4: '0 00:00:00-23:59:59',
            5: '0 00:00:00-23:59:59',
          }),
          4: dict({
            0: '6 00:00:00-23:59:59',
            1: '0 00:00:00-23:59:59',
            2: '0 00:00:00-23:59:59',
            3: '0 00:00:00-23:59:59',
            4: '0 00:00:00-23:59:59',
            5: '0 00:00:00-23:59:59',
          }),
          5: dict({
            0: '6 00:00:00-23:59:59',
            1: '0 00:00:00-23:59:59',
            2: '0 00:00:00-23:59:59',
            3: '0 00:00:00-23:59:59',
            4: '0 00:00:00-23:59:59',
            5: '0 00:00:00-23:59:59',
          }),
          6: dict({
            0: '6 00:00:00-23:59:59',
            1: '0 00:00:00-23:59:59',
            2: '0 00:00:00-23:59:59',
            3: '0 00:00:00-23:59:59',
            4: '0 00:00:00-23:59:59',
            5: '0 00:00:00-23:59:59',
          }),
          7: dict({
            0: '6 00:00:00-23:59:59',
            1: '0 00:00:00-23:59:59',
            2: '0 00:00:00-23:59:59',
            3: '0 00:00:00-23:59:59',
            4: '0 00:00:00-23:59:59',
            5: '0 00:00:00-23:59:59',
          }),
        }),
      }),
    }),
  })
# ---
